### PR TITLE
Improve instance filter input validation

### LINKED ISSUE #177: Improve Instance Filter input validation
Some possible improvements for better UX when entering values in Instance Filter
```[tasklist]
- [x] Validate rule value input `onBlur` or on demand. E.g. some `Apply` button is clicked
- [ ] https://github.com/iTwin/presentation/issues/197
```

### PR DESCRIPTION
**Pull request by simasjar**
closes #177

### CODE REVIEW

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 1]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+    .flat(ruleGroupDepthLimit ?? Number.MAX_SAFE_INTEGER)
+    .some((item) => !isPropertyFilterBuilderRuleGroup(item) && item.operator !== undefined);
 
-  if (!show) {
-    return null;
-  }
   return (
+    <>
+      <Dialog.Content className="presentation-instance-filter-content">
+        <navigationPropertyEditorContext.Provider value={navigationPropertyEditorContextValue}>
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:135]: I think this context should be moved to `InstanceFilterBuilder` to cleanup this component.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 2]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
-    }, 250);
-    return () => {
-      clearTimeout(timeout);
-    };
-  }, []);
+function PresentationInstanceFilterDialogContent(props: PresedntationInstanceFilterDialogContentProps) {
+  const { onApply, initialFilter, descriptor, imodel, ruleGroupDepthLimit, filterResultCountRenderer, onClose } = props;
+
+  const { rootGroup, actions, validate } = usePropertyFilterBuilder({
+    initialFilter: initialFilter ? convertPresentationFilterToPropertyFilter(descriptor, initialFilter.filter) : undefined,
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:111]: We should avoid calling `convertPresentationFilterToPropertyFilter` on each render.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 3]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
-  }
-
-  return <PresentationInstanceFilterBuilder {...restProps} descriptor={descriptor} />;
-}
-
-function useDelayLoadedDescriptor(descriptorGetter: () => Promise<Descriptor>) {
-  const [descriptor, setDescriptor] = useState<Descriptor>();
+function useDelayLoadedDescriptor(descr: Descriptor | (() => Promise<Descriptor>)) {
+  const [descriptor, setDescriptor] = useState<Descriptor | (() => Promise<Descriptor>)>(() => descr);
+  const disposed = useRef<boolean>(false);
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:83]: You could use local variable scoped to `useEffect` for tracking status of disposal.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 4]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
-    return <DelayedCenteredProgressRadial />;
-  }
-
-  return <PresentationInstanceFilterBuilder {...restProps} descriptor={descriptor} />;
-}
-
-function useDelayLoadedDescriptor(descriptorGetter: () => Promise<Descriptor>) {
-  const [descriptor, setDescriptor] = useState<Descriptor>();
+function useDelayLoadedDescriptor(descr: Descriptor | (() => Promise<Descriptor>)) {
+  const [descriptor, setDescriptor] = useState<Descriptor | (() => Promise<Descriptor>)>(() => descr);
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:82]: I would suggest changing type of this to `Descriptor | undefined`. It makes check in component using this function simpler: 

```typescript
const descriptor = useDelayLoadedDescriptor(props.descriptorOrGetter);
if (!descriptor) {
   // render spinner
}
```

File: packages/components/src/presentation-components/instance-filter-builder/Types.ts [hunk 5]
```diff
@@ -48,3 +48,10 @@ export interface PresentationInstanceFilterCondition {
 export function isPresentationInstanceFilterConditionGroup(filter: PresentationInstanceFilter): filter is PresentationInstanceFilterConditionGroup {
   return (filter as any).conditions !== undefined;
 }
+
+export interface PresentationInstanceFilterInfo {
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/Types.ts:52]: Missing comment.

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss [hunk 6]
```diff
@@ -1,12 +1,9 @@
...
-  > input:focus-visible {
-    width: 100%;
-    height: 100%;
-    border: none;
-    text-overflow: ellipsis;
-    overflow: hidden;
-    outline: none;
-    padding: 0;
-  }
+.components-editor-container .iui-input-with-icon .iui-select-button.presentation-navigation-property-select-input > input {
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss:1]: Why all of these additional classes are needed?

File: packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx [hunk 7]
```diff
@@ -48,7 +49,21 @@ describe("InstanceFilter", () => {
...
+  };
+
+  const propertyFilter: PropertyFilter = {
+    property,
+    operator: PropertyFilterRuleOperator.IsNull,
+    value: undefined,
+  };
+
   it("invokes 'onClassSelected' when non selected class is clicked", () => {
+    const { result } = renderHook(() => usePropertyFilterBuilder({ initialFilter: propertyFilter }));
```
Comment (saskliutas) [inline packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx:65]: Can we mock `rootGroup` and `actions` in these tests?

File: packages/components/src/test/instance-filter-builder/PresentationInstanceFilterDialog.test.tsx [hunk 8]
```diff
@@ -118,6 +121,66 @@ describe("PresentationInstanceFilterDialog", () => {
     });
   });
 
+  it("does not invoke `onApply` when `validate` returns undefined", async () => {
```
Comment (saskliutas) [inline packages/components/src/test/instance-filter-builder/PresentationInstanceFilterDialog.test.tsx:124]:
```suggestion
  it("does not invoke `onApply` when filter is invalid", async () => {
```

File: packages/components/src/test/instance-filter-builder/PresentationInstanceFilterDialog.test.tsx [hunk 9]
```diff
@@ -118,6 +121,66 @@ describe("PresentationInstanceFilterDialog", () => {
...
+    // wait until property is selected
+    await waitFor(() => getByDisplayValue(propertiesField.label));
+    expect(applyButton?.disabled).to.be.false;
+
+    fireEvent.click(applyButton!);
+
+    expect(spy).to.not.be.called;
+  });
+
+  it("does not invoke `onApply` when `createPresentationInstanceFilter` returns undefined", async () => {
```
Comment (saskliutas) [inline packages/components/src/test/instance-filter-builder/PresentationInstanceFilterDialog.test.tsx:149]:
```suggestion
  it("does not invoke `onApply` when filter is missing presentation metadata", async () => {
```

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 10]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+        </Dialog.ButtonBar>
+      </div>
+    </>
+  );
+}
+
+function DelayedCenteredProgressRadial() {
+  const show = useDelay();
+
+  return show ? (
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:157]: Why was this changed to ternary operator?

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 11]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+          setDescriptor(newDescriptor);
+        }
       }
     })();
     return () => {
-      disposed = true;
+      disposed.current = true;
     };
-  }, [descriptorGetter]);
+  }, [descr, descriptor]);
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:98]: You shouldn't use `descriptor` state variable inside this `useEffect`

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 12]
```diff
@@ -34,23 +43,16 @@ export interface PresentationInstanceFilterDialogProps extends Omit<Presentation
   filterResultCountRenderer?: (filter?: PresentationInstanceFilterInfo) => React.ReactNode;
   /** Dialog title. */
   title?: React.ReactNode;
+  initialFilter?: PresentationInstanceFilterInfo;
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:46]: Missing comment.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 13]
```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+    if (filter) {
+      const presentationInstanceFilter = createPresentationInstanceFilter(descriptor, filter);
+      if (presentationInstanceFilter) {
+        onApply({ filter: presentationInstanceFilter, usedClasses: filteringProps.selectedClasses });
+      }
+    }
+  };
+
+  const isDisabled = !rootGroup.items
+    .flat(ruleGroupDepthLimit ?? Number.MAX_SAFE_INTEGER)
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:129]: I don't think there is much difference between `flat(ruleGroupDepthLimit)` and `flat(Number.MAX_SAFE_INTEGER)`. So I think this can be simplified to `flat(Number.MAX_SAFE_INTEGER)`

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss (hunk 6 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss:1]: itwinui CSS selectors override this class, so I added them to increase specificity.

Comment (saskliutas) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss:1]: Is `.components-editor-container` really needed? If yes, I would suggest adding some class on a `div` in `NavigationPropertyTargetSelector ` so it works property when used outside `component-react` editors.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 1 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: moved

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 2 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: fixed

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 3 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: fixed

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 4 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:78]: fixed?

File: packages/components/src/presentation-components/instance-filter-builder/Types.ts (hunk 5 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/Types.ts:56]: added

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss (hunk 6 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss]: added `.iui-input` class instead of `.components-editor-container`

File: packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx (hunk 7 above)
Comment (simasjar) [reply packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx]: mocked

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 12 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:46]: added

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 13 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: fixed

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 11 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: `npm run lint` shows an error then
image
added a comment to disable this rule

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss (hunk 6 above)
Comment (saskliutas) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss]: I don't think `.iui-input` is a good solution because that class is added to input element by `@itwin/itwinui-react` and in future versions it might be removed or renamed. I would still suggest adding class on `div` in `NavigationPropertyTargetSelector` component.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 2 above)
Comment (saskliutas) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: `convertPresentationFilterToPropertyFilter` is still called on each render

File: packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx [hunk 14]
```diff
@@ -48,6 +49,20 @@ describe("InstanceFilter", () => {
     sinon.restore();
   });
 
+  const testImodel = {} as IModelConnection;
+  const testDescriptor = {} as Descriptor;
+  const testActions = {} as PropertyFilterBuilderActions;
+
+  const rootGroupMock = moq.Mock.ofType<PropertyFilterBuilderRuleGroup>();
```
Comment (saskliutas) [inline packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx:56]: I think this can be changed to simple object as you are not changing behavior of `rootGroupMock` in any of the tests.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss (hunk 6 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss]: added new div

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 10 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: reverted

File: packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx (hunk 14 above)
Comment (simasjar) [reply packages/components/src/test/instance-filter-builder/IntanceFilterBuilder.test.tsx]: changed

File: packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss [hunk 15]
```diff
@@ -1,12 +1,9 @@
...
-  > input:focus-visible {
-    width: 100%;
-    height: 100%;
-    border: none;
-    text-overflow: ellipsis;
-    overflow: hidden;
-    outline: none;
-    padding: 0;
-  }
+.iui-input-with-icon.presentation-navigation-property-select-control .iui-select-button.presentation-navigation-property-select-input > input {
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss:1]: Would it be enough to specify only `presentation` related css classes?

```suggestion
.presentation-navigation-property-select-control .presentation-navigation-property-select-input > input {
```

Comment (simasjar) [reply packages/components/src/presentation-components/properties/NavigationPropertyTargetSelector.scss]: not really, itwinui selector's specificity is (0,4,0), moved `border: none` and `outline: none` to inline

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 16]
```diff
@@ -66,78 +68,107 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+    if (isPropertyFilterBuilderRuleGroup(item)) {
+      for (const itm of item.items) {
+        if (shouldApplyButtonBeEnabled(itm)) {
+          return true;
+        }
+      }
+      return false;
+    } else {
+      return item.operator !== undefined;
+    }
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:133]:
```suggestion
  const hasNonEmptyRule = (item: PropertyFilterBuilderRuleGroupItem) => {
    if (isPropertyFilterBuilderRuleGroup(item)) {
      return item.items.some(hasNonEmptyRule);
    }
    return item.operator !== undefined;
```

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 17]
```diff
@@ -104,14 +111,36 @@ function PresentationInstanceFilterDialogContent(props: PresedntationInstanceFil
   const { onApply, initialFilter, descriptor, imodel, ruleGroupDepthLimit, filterResultCountRenderer, onClose } = props;
   const [initialPropertyFilter] = useState(() => (initialFilter ? convertPresentationFilterToPropertyFilter(descriptor, initialFilter.filter) : undefined));
 
-  const { rootGroup, actions, validate } = usePropertyFilterBuilder({
+  const numericInputValidator = (item: PropertyFilterBuilderRule) => {
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:114]: I would suggest moving this out of component to reduce noise.

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 18]
```diff
@@ -104,14 +111,36 @@ function PresentationInstanceFilterDialogContent(props: PresedntationInstanceFil
...
+      item.value.value === undefined &&
+      item.value.displayValue !== ""
+    ) {
+      return translate("instance-filter-builder.error-messages.notANumber");
+    }
+    let errorMessage = defaultPropertyFilterBuilderRuleValidator(item);
+    if (errorMessage) {
+      return errorMessage;
+    }
+    return undefined;
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:132]:
```suggestion
    return defaultPropertyFilterBuilderRuleValidator(item);
```

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 19]
```diff
@@ -104,14 +111,36 @@ function PresentationInstanceFilterDialogContent(props: PresedntationInstanceFil
...
+    if (
+      item.value &&
+      item.value.valueFormat === PropertyValueFormat.Primitive &&
+      item.property &&
+      (item.property.typename === StandardTypeNames.Number ||
+        item.property.typename === StandardTypeNames.Int ||
+        item.property.typename === StandardTypeNames.Float ||
+        item.property.typename === StandardTypeNames.Double) &&
+      item.value.value === undefined &&
+      item.value.displayValue !== ""
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:124]: Would it be possible to reduce conditions in this `if`? Maybe move some to separate functions (like checking if property is numeric)

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx (hunk 19 above)
Comment (simasjar) [reply packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx]: moved checking if property is numeric to other function

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 20]
```diff
@@ -68,81 +77,126 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
...
+  const filteringProps = usePresentationInstanceFilteringProps(descriptor, imodel, initialFilter?.usedClasses);
+
+  const applyButtonHandle = () => {
+    const filter = buildFilter();
+    if (filter) {
+      const presentationInstanceFilter = createPresentationInstanceFilter(descriptor, filter);
+      if (presentationInstanceFilter) {
+        onApply({ filter: presentationInstanceFilter, usedClasses: filteringProps.selectedClasses });
+      }
+    }
```
Comment (saskliutas) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:130]:
```suggestion
    const filter = buildFilter();
    if (!filter) {
      return;
    }
    const presentationInstanceFilter = createPresentationInstanceFilter(descriptor, filter);
    if (!presentationInstanceFilter) {
      return;
    }
    onApply({ filter: presentationInstanceFilter, usedClasses: filteringProps.selectedClasses });
```

File: packages/components/public/locales/en/PresentationComponents.json [hunk 21]
```diff
@@ -21,7 +21,10 @@
     "schema": "Schema:",
     "apply": "Apply",
     "cancel": "Cancel",
-    "filter": "Filter"
+    "filter": "Filter",
+    "error-messages": {
+      "notANumber": "Value is not a number"
```
Comment (saskliutas) [inline packages/components/public/locales/en/PresentationComponents.json:26]:
```suggestion
      "notANumber": "Value must be a number"
```

**Review by saskliutas — APPROVED** [review]
